MCH: properly configure the slot finalization

aferrero2707 opened the pull request:
The call to `setSlotLength(o2::calibration::INFINITE_TF)` does not seem to be appropriate in order to trigger the slot finalization at EoS, because it does not set the `mFinalizeWhenReady` data member of the `TimeSlotCalibration` base class. This is needed in order to properly trigger the slot finalization when calling `checkSlotsToFinalize(o2::calibration::INFINITE_TF)`.

An explicit call to `setFinalizeWhenReady()` is instead needed. Internally this also calls `setSlotLength(o2::calibration::INFINITE_TF)`.

aferrero2707 commented:
@aphecetche this PR is expected to fix the issues we are having with the empty bad channels calibrations.

@chiarazampolli @shahor02 could you please have a look and confirm that the fix is correct?
There is also something I do not fully understand in the [TimeSlotCalibration](https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h) code, in particular in the calls to [setSlotLength()](https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h#L54) and [setFinalizeWhenReady()](https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h#L92): the two functions seem to call each other, but with a different logic regarding the value of the `TFType` parameter.

shahor02 commented:
@aferrero2707, 
(1) with  `setFinalizeWhenReady` you tell to calibrator to finalize the slot and to stop the processing once wanted stat. is accumulated (w/o waiting for the EoR). It forces the slot length to infinity just to make sure that everything goes to 1 slot.
(2) In opposite, simply setting the slot length to infinity means that the slot will be filled for every TF w/o attempting to be finalized. 
The finalization will happen at the endOfStream call of the calibrator.

If the behaviour you want is (1), then the your PR is fine.

`setSlotLength(0)`, which calls the  `setFinalizeWhenReady` is just the way to set this option via dummy slot length.

aferrero2707 commented:
@shahor02 we actually want to keep both behaviors, although the processing at End-of-Stream is the preferred one. However, we want the possibility to create the calibration objects once a given statistics is reached, in case the EoS handling is broken for some reason.

However, when `mFinalizeWhenReady = false` and `checkSlotsToFinalize()` is called with `tf = INFINITE_TF`, as we are doing [here](https://github.com/AliceMCH/AliceO2/blob/mrrtf-204-fix-eos-in-calibrator/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx#L160), the slot finalization never happens because [this check](https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h#L317) is always false (`tf` and `tfLim` are both equal to `INFINITE_TF`). Could this be a bug?

At this point I have the impression that we should modify/simply our calibrator logic:
* we should call `setUpdateAtTheEndOfRunOnly()` only when using the EoS method. In this case we should not call `checkSlotsToFinalize()` in our calibrator code, but instead directly call `finalizeSlot()` passing the one and only active slot (since the slot length is set to `INFINITE_TF`)
* when we do not rely on EoS and we require a minimum statistics instead, we should not call `setUpdateAtTheEndOfRunOnly()` and let the calibrator's `process()` function handle the slot finalization

Does this make sense?

aferrero2707 commented:
@shahor02 one more question, as I am still trying to understand the details of the underlying logic:

> with setFinalizeWhenReady you tell to calibrator to finalize the slot and to stop the processing once wanted stat. is accumulated (w/o waiting for the EoR).

How the processing is supposed to be stopped? Should it be manually stopped in the calibrator device code, or is it supposed to happen automatically?

As far as I can see, if the processing is not explicitly stopped in the DPL device code, then a new slot is inserted and starts to be filled...

shahor02 commented:
@aferrero2707 sorry, rechecking the code I see that what I wrote before is wrong: the FinalizeWhenReady first defines a slot with an infinite end of validity and once this slot gets sufficient statistics it is finalized, its end of validity is set to the highest TF seen so far. Then a new slot is defined, from where the previous has ended till `INFINITE_TF` and the same processing continues. Untill the endOfStream calls the `checkSlotsToFinalize` with `TF = INFINITE_TF`, meaning that the processing is over.
So, 
1) if you want to have multiple calibration slots within the run with the boundaries defined according to the collected statistics, then the setFinalizeWhenReady is right way to go. 

2) If you want calibration slots of fixed duration, you just set setSlotLength (or setSlotLengthInSeconds or setSlotLengthInOrbits).

3) If you want to process the data of the whole run in a single calib. slot which is finalized at the endOfStream, then you should  `setSlotLength(o2::calibration::INFINITE_TF-1)`. This will ensure finalization in the endOfStream call of `checkSlotsToFinalize(o2::calibration::INFINITE_TF)` (or simply `checkSlotsToFinalize()`).
Note that in this case, you can set `setUpdateAtTheEndOfRunOnly()` to not spend time on the checks for slots to finalize in the run(..) method.

aferrero2707 commented:
@shahor02 in our case we want to process the data of the whole run in a single calib. slot, which is finalized either at the EndOfStream or as soon as a target statistics is reached (the two modes are controlled by the `BadChannelCalibratorParam::Instance().onlyAtEndOfStream` configuration option).

Given that we only have one slot per run, and that we were already explicitly checking the slot statistics in the `run()` method, I considered that we do not really need the complexity of the `checkSlotsToFinalize()` function, and that we can easily handle the slot finalization explicitly.

Hence I modified/simplified the code in the following way:
1. in the `run()` method, we check the value of the `onlyAtEndOfStream` flag (see [here](https://github.com/AliceMCH/AliceO2/blob/mrrtf-204-fix-eos-in-calibrator/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx#L87)):
    * if `false` we check the slot statistics ([here](https://github.com/AliceMCH/AliceO2/blob/mrrtf-204-fix-eos-in-calibrator/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx#L89)) and we trigger the slot finalization if the statistics is sufficient (see [here](https://github.com/AliceMCH/AliceO2/blob/mrrtf-204-fix-eos-in-calibrator/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx#L174) and [here](https://github.com/AliceMCH/AliceO2/blob/mrrtf-204-fix-eos-in-calibrator/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrator.cxx#L54) for the details). Further data is then skipped once the slot is finalized and the CCDB populated.
    * if `true` the statistics is not checked in the `run()` function and everything is postponed to the EndOfStream
2. in the `sendOutput()` method we directly call the slot finalization (through the [BadChannelCalibrator::finalize(https://github.com/AliceMCH/AliceO2/blob/mrrtf-204-fix-eos-in-calibrator/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrator.cxx#L54)]() helper method) instead of going through the `checkSlotsToFinalize()`, because we only have one single slot to finalize and there is no need to start a new one afterward

Do you have objections to this approach, or suggestions to improve the code?

Thanks!

aphecetche commented on Detectors/MUON/MCH/Calibration/include/MCHCalibration/PedestalChannel.h:
```diff
@@ -16,18 +16,19 @@
+  double mPedestal{0};     ///< mean
+  double mVariance{0};     ///< variance
+  DsChannelId dsChannelId; ///< identifier of the channel
+  bool mValid{false};      ///< true if the channel is associated to a detector pad
```
do we really need an extra flag here ? Can't a test on DsChannelId  give the same information ?

aferrero2707 replied:
Agreed. I have replaced the `mValid` member by a `isValid()` function that actually checks the value of the `dsChannelId`.

aphecetche commented on Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx:
```diff
@@ -35,8 +35,8 @@ void BadChannelCalibrationDevice::init(o2::framework::InitContext& ic)
   mCalibrator = std::make_unique<o2::mch::calibration::BadChannelCalibrator>();
-  mCalibrator->setSlotLength(o2::calibration::INFINITE_TF);
   mCalibrator->setUpdateAtTheEndOfRunOnly();
+  mCalibrator->setSlotLength(o2::calibration::INFINITE_TF);
```
here you changed the order the calls to `setUpdateAtTheEndOfRunOnly` and `setSlotLength` : is it on purpose ? in which case would be good to add a comment right there in the code to say so.
If not, just forget this comment ;-)

shahor02 replied:
The order should not matter, but you should set 
```
setSlotLength(o2::calibration::INFINITE_TF);
setMaxSlotsDelay(0);
```
to guarantee that the endOfStream call with `tf = INFINITE_TF`) passes the https://github.com/AliceO2Group/AliceO2/blob/b0f70e90f824acf876f63ca28f8ddeffb6b9523b/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h#L315-L317 check.

aferrero2707 replied:
No, it was not done on purpose, reverted.

shahor02 replied:
@aferrero2707 a typo slipped to my previous comment, should be 
```
setSlotLength(o2::calibration::INFINITE_TF - 1);
setMaxSlotsDelay(0);
```
to guarantee that the endOfStream call with tf = INFINITE_TF) passes the https://github.com/AliceO2Group/AliceO2/blob/b0f70e90f824acf876f63ca28f8ddeffb6b9523b/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h#L315-L317

aferrero2707 replied:
@shahor02 I have to confess that I still do not fully understand the logic.. 
Calling `setMaxSlotsDelay(0)` will set `mFinalizeWhenReady` to `true`. If I understand correctly, this means that the program will not go through the lines you quote, instead it will reach this portion of code:
https://github.com/AliceO2Group/AliceO2/blob/b0f70e90f824acf876f63ca28f8ddeffb6b9523b/Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h#L279-L286
In this case, the slot is indeed finalized as soon as the desired statistics is reached, but then the slot is removed and a new one is inserted. However, we do not really need this new slot insertion, because we only use one slot for the full run...

shahor02 replied:
> Calling `setMaxSlotsDelay(0)` will set `mFinalizeWhenReady` to `true`. 

No, you are confusing it with `setSlotLength(0)` behaviour,  the mMaxSlotsDelay (the unit is a fraction of the slot length) defines maximum difference `tf_current - slot[I]_end_tf` after which the slot `#I` is allowed to be checked for finalization (i.e. no more TFs contributing to it are expected). 
Though after rechecking the code I see that it will work also w/o `setMaxSlotsDelay(0)` setting, the important part is  `setSlotLength(INFINITE_TF-1)`:
since the endOfStream calls  `checkSlotsToFinalize` with `tf==INFINITE_TF`, and the slot can be finalized only if 
`"current" tf >  slot->getTFEnd()`, the `slot->getTFEnd()` must be `<INFINITE_TF`.

aferrero2707 replied:
@shahor02 my bad, I completely mixed things up! You are absolutely correct, and `setSlotLength(INFINITE_TF-1)` is ok to trigger the slot finalization.

Nevertheless, I would still be in favor of not using the `checkSlotsToFinalize()` function in our code, for the following reasons:
* we initialize the calibrator with `setUpdateAtTheEndOfRunOnly()`, therefore we do not take advantage of the `checkSlotsToFinalize()` call at each TF in the data processing routine
* we anyhow check explicitly the available statistics in the `run()` method (see [here](https://github.com/AliceMCH/AliceO2/blob/9008f679c2ca05695ecbd439dad9d8d6db2ed413/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx#L87)) when we do not rely on the EoS
* when the EoS is enabled, we finalize the slot even if the statistics is not enough, because we want to always send the output to the QC for diagnostics purposes - AFAIK this is not possible with the `checkSlotsToFinalize()` method

Do you see any drawbacks?

shahor02 replied:
@aferrero2707 no, I don't see drawbacks.

aphecetche commented on Detectors/MUON/MCH/Calibration/src/BadChannelCalibrationDevice.cxx:
```diff
@@ -72,16 +72,26 @@ void BadChannelCalibrationDevice::logStats(size_t dataSize)
 
 void BadChannelCalibrationDevice::run(o2::framework::ProcessingContext& pc)
 {
+  static bool skip = false;
```
This might not matter much, but the usage of this static means that the same ECS environment can *not* be reused to perform two pedestal runs, right ?

aferrero2707 replied:
Good point! I have replaced the static variable by a member function, to avoid leaving some herd-to-debug checks in the code.

aphecetche commented on Detectors/MUON/MCH/Calibration/src/BadChannelCalibrator.cxx:
```diff
@@ -34,14 +34,12 @@ void BadChannelCalibrator::initOutput()
 
   // let's check our hypothesis about this object (nslots=1) is actually true
   auto nslots = getNSlots();
   if (nslots != 1) {
```
I'm not sure we still need this check here as it is now duplicated in the `finalize` method which is anyways called, right ?

aferrero2707 replied:
I propose to keep it in order to make sure that the subsequent call to `hasEnoughData()` takes the expected slot as input. In any case this is a fatal condition that we should never encounter.

aphecetche commented on Detectors/MUON/MCH/Calibration/src/BadChannelCalibrator.cxx:
```diff
@@ -61,14 +71,13 @@ bool BadChannelCalibrator::hasEnoughData(const Slot& slot) const
 
   auto nofCalibrated = std::count_if(pedData->cbegin(), pedData->cend(),
-                                     [&](const PedestalChannel& c) { return c.mEntries > minNofEntries; });
+                                     [&](const PedestalChannel& c) { return (c.mValid && (c.mEntries > minNofEntries)); });
```
actually, instead of testing for `mValid` (see also the earlier comment about whether or not we really need this extra boolean) here and in line 103, the skipping of the non existing channel could be baked in the iterator directly.

aferrero2707 replied:
I am not sure to see how to implement this in the iterator... As explained above, I have instead replaced the `mValid` member by a `isValid()` function that actually checks the value of the `dsChannelId`.

aphecetche replied:
for the iterator, see https://github.com/AliceMCH/AliceO2/pull/34

aferrero2707 committed:
[MCH] improve logic of slot finalization

The check of the slot statistics and the slot finalization are explicitly
done in the calibrator code instead of relying on the `checkSlotsToFinalize()`
function, because for MCH we only have one slot that spans the whole run.

`setUpdateAtTheEndOfRunOnly()` is also called at initialization to skip the
internal check of the slot statistics at each TF.

In the case where the calibrations are produced as soon as a target statistics
is reached, the subsequent data is skipped to avoid producing multiple
calibration objects for the same run.

aferrero2707 committed:
[MCH] use mapping in pedestals calibration

The electronics mapping is used to flag as valid the readout channels that
are associated to detector pads. Only valid channels are processed and
used to compute the total pedestals data size. Invalid channels therefore
will not contribute to the computation when checking the statistics of
a given calibration slot.

aferrero2707 committed:
[MCH] fix clang formatting errors

alibuild commented:
Error while checking build/O2/fullCI for 0ba2f884b61aab6edd5bd9a25614ee0a4bb9131f at 2023-01-19 01:36:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10566/0ba2f884b61aab6edd5bd9a25614ee0a4bb9131f/build_O2_fullCI/pretty.html).

aphecetche committed:
MCH: update ped calib example script

aphecetche committed:
MCH: pedestal iterator only iterates on valid channels

aphecetche committed:
MCH: update testPedestalData test

alibuild commented:
Error while checking build/O2/fullCI for 9008f679c2ca05695ecbd439dad9d8d6db2ed413 at 2023-01-19 23:37:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10566/9008f679c2ca05695ecbd439dad9d8d6db2ed413/build_O2_fullCI/pretty.html).

aphecetche commented:
@shahor02 @chiarazampolli as far as I can tell the failure has nothing to do with this PR. Could it be merged so we'll be able to test it at Pt2 later on ?
